models/src/main/scala/coop/rchain/models/SortedParHashSet.scala

```diff
@@ -1,41 +1,32 @@
 package coop.rchain.models
 
-import scala.collection.generic.CanBuildFrom
-import scala.collection.immutable.HashSet
-import scala.collection.{mutable, SetLike}
 import coop.rchain.models.rholang.sorter.ordering._
 
+import scala.collection.SetLike
+import scala.collection.immutable.HashSet
+
 //Enforce ordering and uniqueness.
 // - uniqueness is handled by using HashSet.
 // - ordering comes from sorting the elements prior to serializing.
-case class SortedParHashSet(private val ps: HashSet[Par])
+class SortedParHashSet(private val ps: HashSet[Par])
     extends Set[Par]
     with SetLike[Par, SortedParHashSet] {
 
   lazy val sortedPars: List[Par] = ps.toList.sort
 
-  override def equals(that: Any): Boolean = that match {
-    case shs: SortedParHashSet => shs.sortedPars == this.sortedPars
-    case _                     => false
-  }
+  override def empty: SortedParHashSet = new SortedParHashSet(HashSet.empty[Par])
 
-  override def hashCode(): Int = sortedPars.hashCode()
+  override def contains(elem: Par): Boolean = sortedPars.contains(elem)
 
-  override def empty: SortedParHashSet = SortedParHashSet.empty
+  override def +(elem: Par): SortedParHashSet = new SortedParHashSet(ps + elem)
 
-  override def contains(elem: Par): Boolean = ps.contains(elem)
-
-  override def +(elem: Par): SortedParHashSet = SortedParHashSet(ps + elem)
-
-  override def -(elem: Par): SortedParHashSet = SortedParHashSet(ps - elem)
+  override def -(elem: Par): SortedParHashSet = new SortedParHashSet(ps - elem)
```

Comment:
There's another bug lurking in here. Both current and the `sortedPs`-based implementations have  the problem, that they can only remove an element with the same sorted-ness (and exact same unsorted shape if we're talking about the current impl).

We should be using `sortedPs` and sort the input `elem: Par` before attempting to remove it. This will work whether `elem` was originally sorted or not.

This ofc is a problem for the Map too... But let's fix that in a follow-up PR after this and #1729 gets merged.

Reply:
How would the lazy init blow up with an NPE?

Reply:
I think I've mixed val initialization order problems and implicit initialization problems with `lazy val` (which seems to be brought up more often as a solution than a problem). Scala is "funny" language ;/... Please disregard the mentioned comment :)